### PR TITLE
feat(symphony): add symphony doctor preflight diagnostics

### PR DESCRIPTION
**Pull request by gannonh**
## Summary
- add a new `symphony doctor [WORKFLOW.md]` subcommand with deterministic traffic-light output and 0/1 exit behavior
- implement doctor checks for config parsing/validation, env var references, prompt file paths, Linear connectivity, backend command handshake, workspace health, and orphan workspace detection
- add doctor-focused test coverage (unit + CLI integration), including mock-backed full-run coverage and auth-failure cases
- document doctor usage/check categories/exit codes in `apps/symphony/README.md` and `apps/symphony/docs/WORKFLOW-REFERENCE.md`

## Validation
- `cd apps/symphony && cargo test --test cli_tests doctor`
- `cd apps/symphony && cargo test doctor`
- `cd apps/symphony && cargo test`
- `cd apps/symphony && cargo clippy -- -D warnings`
- Manual: `cargo run --quiet -- doctor <temp WORKFLOW.md with invalid api_key>` (verified `🚨 Linear Auth ... HTTP 401`, exit code 1)

Closes KAT-1675

<h3>Greptile Summary</h3>

This PR introduces a `symphony doctor [WORKFLOW.md]` preflight diagnostics subcommand that runs six groups of checks (config parse/validate, env-var resolution, Linear auth/project/states/assignee, backend handshake, workspace health, orphan detection) and exits 0/1 based on whether any 🚨 errors are found. The implementation is well-structured with clean separation between the pure check functions in `doctor.rs` and the async-dispatch glue in `run_doctor` in `main.rs`. Test coverage is solid, including a mock-backed full happy-path run and auth-failure cases.

Key findings:

- **`run_doctor` implicit tokio runtime dependency** (`main.rs` lines 551-555): `Handle::try_current()` + `block_in_place` silently fails if no active runtime is present or **panics** if the active runtime is `current_thread` flavored. Creating a dedicated runtime inside `run_doctor` would make it self-contained.
- **Probe-file leak in `assert_directory_writable`** (`doctor.rs` lines 1145-1146): If `fs::write` succeeds but `fs::remove_file` fails, a `.symphony-doctor-write-probe-*` file is left in `workspace.root`.
- **`${VAR}` env-var syntax not detected** (`doctor.rs` lines 1087-1099): `env_reference_name` only matches the `$VAR` bare-dollar form; users who write `${LINEAR_API_KEY}` will not receive an unresolved-variable warning.
- **Documentation inconsistency** (`doctor.rs` line 565, `WORKFLOW-REFERENCE.md`): `check_workspace` creates the workspace root directory as a side effect, but `WORKFLOW-REFERENCE.md` states \"Doctor is report-only (no auto-fix).\"
- **Workflow file parsed twice** (`main.rs` lines 546-550): `check_config` and `load_service_config` each independently parse the workflow file.

<h3>Confidence Score: 5/5</h3>

Safe to merge; all findings are P2 style/improvement suggestions that do not block functionality.

All five findings are P2 level — none represent a present runtime breakage or data-loss risk. The block_in_place concern is real but the manual test validated it works in the actual runtime topology. The probe-file leak and ${VAR} gap are edge cases. The double-parse is a minor efficiency concern. The doc inconsistency is cosmetic. The new feature is well-tested with mocks and integrates cleanly with the existing CLI structure.

apps/symphony/src/doctor.rs (assert_directory_writable probe leak, env_reference_name ${VAR} gap), apps/symphony/src/main.rs (run_doctor runtime dependency), apps/symphony/docs/WORKFLOW-REFERENCE.md (no-auto-fix claim)

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/symphony/src/doctor.rs | New 1220-line module implementing all doctor check functions (config, linear, backend, workspace, orphans); probe-file leak in assert_directory_writable and ${VAR} env-ref gap are the main issues. |
| apps/symphony/src/main.rs | Adds CliCommand::Doctor subcommand, run_doctor orchestrator, and run_entrypoint dispatch; run_doctor has an implicit tokio runtime dependency (block_in_place) and double-parses the workflow file. |
| apps/symphony/tests/cli_tests.rs | Adds comprehensive doctor test coverage: CLI parse, missing API key, non-existent backend, workspace strategy mismatch, auth failure via mockito, and a full mock-backed happy-path run. |
| apps/symphony/src/lib.rs | Single-line addition exposing the new doctor module publicly; no issues. |
| apps/symphony/README.md | Documents doctor subcommand usage, traffic-light symbols, example output, check categories, and exit codes; clear and accurate. |
| apps/symphony/docs/WORKFLOW-REFERENCE.md | Adds doctor CLI reference comment block to the WORKFLOW.md template; "report-only (no auto-fix)" claim is contradicted by check_workspace's directory-creation side effect. |

</details>

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A([symphony doctor WORKFLOW.md]) --> B[check_config]
    B --> B1{Parse YAML}
    B1 -- Error --> BAIL1[🚨 Config Parse\nreturn early]
    B1 -- OK --> B2[collect_unresolved_env_refs]
    B2 --> B3[collect_invalid_slack_events\nsanitize_invalid_slack_events]
    B3 --> B4[config::from_workflow\nconfig::validate]
    B4 --> B5[collect_prompt_paths\ncheck file existence]
    B5 --> HAS_ERR{has_errors?}
    HAS_ERR -- Yes --> SKIP[⏭️ Runtime Checks Skipped]
    HAS_ERR -- No --> LC[load_service_config]
    LC --> CL[check_linear async]
    CL --> CL1[VIEWER_QUERY Linear Auth]
    CL1 --> CL2[PROJECT_QUERY Linear Project + States]
    CL2 --> CL3[resolve_assignee paginated]
    LC --> CB[check_backend sync]
    CB --> CB1[executable --version]
    LC --> CW[check_workspace sync]
    CW --> CW1[fs::create_dir_all if missing]
    CW1 --> CW2[repo_is_remote + git_strategy compat]
    CW2 --> CW3[docker info if isolation=docker]
    LC --> CO[check_orphans async]
    CO --> CO1[scan_workspace_root]
    CO1 --> CO2[fetch_issues_by_states]
    CO2 --> CO3[diff on-disk vs active]
    CL --> FR
    CB --> FR
    CW --> FR
    CO --> FR
    SKIP --> FR
    FR[format_results print to stdout]
    FR --> EXIT{has_errors?}
    EXIT -- Yes --> E1([exit 1])
    EXIT -- No --> E0([exit 0])
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: apps/symphony/src/doctor.rs
Line: 1145-1147

Comment:
**Probe file left on disk if removal fails**

`fs::write` creates the probe file, but if `fs::remove_file` then fails (e.g., due to a concurrent directory rename, a file-system error, or a signal), the function propagates the error while leaving the `.symphony-doctor-write-probe-*` file behind in `workspace.root`. The caller treats the I/O error as "not writable" and reports an error, but the probe file persists until manually cleaned up.

A safer pattern is to use a guard / `defer` equivalent, or at minimum attempt the removal even on error paths:

```suggestion
    fs::write(&probe_path, b"doctor")?;
    let remove_result = fs::remove_file(&probe_path);
    remove_result?;
    Ok(())
```

Alternatively, clean up proactively even when removal fails:
```rust
fs::write(&probe_path, b"doctor")?;
fs::remove_file(&probe_path).ok(); // best-effort removal; not a write-ability signal
Ok(())
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/symphony/src/doctor.rs
Line: 1087-1099

Comment:
**`${VAR}` env-var syntax silently skipped**

`env_reference_name` only recognises the `$VAR` bare-dollar form. The very common shell expansion `${LINEAR_API_KEY}` (curly-brace delimited) starts with `$` but the remainder contains `{` so `is_bare_identifier` would include it (since `{` is not filtered), yet the resolved value would be the literal string `{LINEAR_API_KEY}` rather than the env var content — meaning a genuinely unset `${VAR}` will _not_ be flagged.

Consider expanding the check to strip `{` and `}` delimiters:

```rust
fn env_reference_name(value: &str) -> Option<&str> {
    let variable = value.strip_prefix('$')?;
    let variable = variable
        .strip_prefix('{')
        .and_then(|v| v.strip_suffix('}'))
        .unwrap_or(variable);
    let is_bare_identifier = !variable.is_empty()
        && !variable.contains('/')
        && !variable.contains(' ')
        && !variable.contains(':');
    if is_bare_identifier { Some(variable) } else { None }
}
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/symphony/src/doctor.rs
Line: 565-579

Comment:
**Directory creation contradicts "no auto-fix" documentation**

`check_workspace` calls `fs::create_dir_all` when `workspace.root` does not exist and reports a ✅ pass with "Created workspace.root: ...". The `README.md` example output implies this is intentional, but `WORKFLOW-REFERENCE.md` states:

> `Doctor is report-only (no auto-fix).`

Silently creating a directory is a side-effectful mutation that could be surprising to users running `doctor` in CI expecting a pure no-op validation.

Consider either:
1. Changing the behaviour to report ⚠️ warning "workspace.root does not exist; it will be created on first run" and skip the write-probe in that path, OR
2. Updating `WORKFLOW-REFERENCE.md` to clarify that `doctor` may create `workspace.root` as part of its validation.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/symphony/src/main.rs
Line: 551-555

Comment:
**`block_in_place` panics on `current_thread` runtime flavor**

`run_doctor` depends on two implicit assumptions:

1. A tokio runtime is already active when it is called (`Handle::try_current()`). If the runtime is created _inside_ `execute_cli` (after the doctor dispatch), `try_current()` fails and doctor exits with "missing tokio runtime for doctor checks" instead of running any async checks.
2. The active runtime is multi-threaded. `tokio::task::block_in_place` **panics** at runtime if called from a `current_thread` runtime (e.g. `#[tokio::main(flavor = "current_thread")]`).

A self-contained approach avoids both hazards:

```rust
let rt = tokio::runtime::Builder::new_multi_thread()
    .enable_all()
    .build()
    .map_err(|err| format!("failed to build tokio runtime: {err}"))?;

let linear_results = rt.block_on(doctor::check_linear(&service_config.tracker));
let orphan_results = rt.block_on(doctor::check_orphans(&service_config, &adapter));
```

This keeps `run_doctor` independent of the surrounding runtime topology.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/symphony/src/main.rs
Line: 546-550

Comment:
**Workflow file parsed twice per `doctor` invocation**

`run_doctor` first calls `doctor::check_config(workflow_path)` (which internally calls `workflow::parse_workflow` and `config::from_workflow`), and then — if no errors — calls `doctor::load_service_config(workflow_path)` which repeats the same parse-and-sanitize sequence to produce the `ServiceConfig` needed for runtime checks.

A cleaner design would be to have `check_config` return the parsed `ServiceConfig` alongside its results (e.g., `(Vec<DoctorCheckResult>, Option<ServiceConfig>)`), so `run_doctor` can reuse it directly. This avoids redundant I/O and parsing in the hot path.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["Merge remote-tracking branch &#39;origin/mai..."](https://github.com/gannonh/kata/commit/ca30fb51e51972de551d9ac554235b659443f5e3) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=26677053)</sub>

> Greptile also left **5 inline comments** on this PR.

### CODE REVIEW

File: apps/symphony/src/doctor.rs [hunk 1]
```diff
@@ -0,0 +1,1220 @@
...
+                    notifications::SUPPORTED_SLACK_EVENTS.join(", ")
+                ),
+            ));
+        }
+    }
+
+    let sanitized_config = sanitize_invalid_slack_events(&workflow_definition.config);
+
+    let service_config = match config::from_workflow(&sanitized_config) {
+        Ok(config) => config,
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/symphony/src/doctor.rs]: Fixed in `94ed2f65`/`00266e1c`.

`doctor::check_config` and `doctor::load_service_config` now validate the raw workflow config (no Slack-event sanitization before `config::from_workflow`). Invalid `notifications.slack.events` therefore remain fatal just like normal startup. Added regression coverage in `test_check_config_invalid_slack_event_is_fatal`.

File: apps/symphony/src/doctor.rs [hunk 2]
```diff
@@ -0,0 +1,1220 @@
...
+            "Config Env",
+            "All $ENV_VAR references resolved",
+        ));
+    } else {
+        for issue in unresolved_env_refs {
+            results.push(DoctorCheckResult::error(
+                "Config Env",
+                format!(
+                    "{} references ${} but the environment variable is unset or empty",
+                    issue.path, issue.var_name
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/symphony/src/doctor.rs]: Fixed in `00266e1c`.

Unresolved env refs are no longer uniformly fatal: doctor now treats required tracker refs (`tracker.api_key`, `tracker.project_slug`) as 🚨 and optional refs as ⚠️ so optional `$VAR` usage (e.g. `tracker.assignee`) doesn’t incorrectly fail the run. Added regression test `test_check_config_optional_env_reference_is_warning`.

File: apps/symphony/src/doctor.rs [hunk 3]
```diff
@@ -0,0 +1,1220 @@
...
+fn assert_directory_writable(path: &Path) -> std::io::Result<()> {
+    let probe_path = path.join(format!(
+        ".symphony-doctor-write-probe-{}-{}",
+        std::process::id(),
+        chrono::Utc::now().timestamp_nanos_opt().unwrap_or_default()
+    ));
+
+    fs::write(&probe_path, b"doctor")?;
+    fs::remove_file(probe_path)?;
+    Ok(())
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/symphony/src/doctor.rs:1119]: Addressed in `94ed2f65`.

`assert_directory_writable` now uses a drop guard (`ProbeCleanup`) so probe-file cleanup is attempted even on error paths, eliminating stale probe files from the common failure mode where `remove_file` errors.

File: apps/symphony/src/doctor.rs [hunk 4]
```diff
@@ -0,0 +1,1220 @@
...
+                        format!(
+                            "workspace.root was created but is not writable ({}): {err}",
+                            workspace_root.display()
+                        ),
+                    ));
+                } else {
+                    results.push(DoctorCheckResult::pass(
+                        "Workspace Root",
+                        format!("Created workspace.root: {}", workspace_root.display()),
+                    ));
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/symphony/src/doctor.rs:579]: Good catch — clarified in docs in `94ed2f65`.

I kept the existing behavior (`workspace.root` may be created to validate writability), but updated both references to make that explicit:
- `apps/symphony/README.md`
- `apps/symphony/docs/WORKFLOW-REFERENCE.md`

So the no-auto-fix claim now documents the intentional local `workspace.root` creation exception.

File: apps/symphony/src/doctor.rs [hunk 5]
```diff
@@ -0,0 +1,1220 @@
...
+        && !variable.contains('/')
+        && !variable.contains(' ')
+        && !variable.contains(':');
+
+    if is_bare_identifier {
+        Some(variable)
+    } else {
+        None
+    }
+}
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/symphony/src/doctor.rs:1061]: I’m leaving this as-is intentionally.

Symphony’s workflow contract currently supports **bare** `$VAR` references only (documented in WORKFLOW reference and mirrored in `config::env_reference_name`). `${VAR}` is not a supported syntax in runtime config parsing today.

If we teach doctor-only parsing to accept `${VAR}`, doctor would diverge from startup behavior. I’d rather keep parity and address `${VAR}` support holistically (config + doctor + docs) in a dedicated follow-up change.

File: apps/symphony/src/main.rs [hunk 6]
```diff
@@ -523,6 +541,55 @@ pub fn execute_cli(cli: &Cli, deps: &mut dyn BootstrapDeps) -> Result<(), String
...
+    let mut results = doctor::check_config(workflow_path);
+
+    if !doctor::has_errors(&results) {
+        match doctor::load_service_config(workflow_path) {
+            Ok(service_config) => {
+                let runtime = tokio::runtime::Handle::try_current()
+                    .map_err(|err| format!("missing tokio runtime for doctor checks: {err}"))?;
+
+                let linear_results = tokio::task::block_in_place(|| {
+                    runtime.block_on(doctor::check_linear(&service_config.tracker))
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/symphony/src/main.rs:555]: Keeping this implementation for now.

`run_doctor` is only called from the production binary path under `#[tokio::main]`, which uses the multi-thread runtime flavor by default. In that runtime, `Handle::try_current()` + `block_in_place` is valid and avoids introducing a second runtime instance.

I agree this would need refactoring if we ever expose doctor from a `current_thread` runtime entrypoint, but that’s not a supported topology for this binary today.

File: apps/symphony/src/main.rs [hunk 7]
```diff
@@ -523,6 +541,55 @@ pub fn execute_cli(cli: &Cli, deps: &mut dyn BootstrapDeps) -> Result<(), String
     })
 }
 
+#[cfg(not(test))]
+fn run_doctor(workflow_path: &Path) -> Result<i32, String> {
+    let mut results = doctor::check_config(workflow_path);
+
+    if !doctor::has_errors(&results) {
+        match doctor::load_service_config(workflow_path) {
+            Ok(service_config) => {
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/symphony/src/main.rs:550]: Good optimization idea; I’m deferring it for now.

This path is a one-shot CLI preflight command, so the extra parse is negligible versus the network/disk checks. Returning `(results, Option<ServiceConfig>)` from `check_config` would also widen the doctor API surface and force call-site/test churn. I’d rather keep the simpler API in this slice and revisit if we see measurable latency pressure.
